browser/ui/views/brave_actions/brave_actions_container.h

```diff
@@ -0,0 +1,105 @@
+#include "extensions/common/extension.h"
+
+namespace extensions {
+  class ExtensionActionManager;
+  class ExtensionRegistry;
+  class ExtensionRegistryObserver;
+}
+
+// TODO: split to separate model, like ToolbarActionsModel and ToolbarActionsBar
+class BraveActionsContainer : public views::View,
+                             public extensions::ExtensionActionAPI::Observer,
+                             public extensions::ExtensionRegistryObserver,
+                             public ToolbarActionView::Delegate {
+  public:
+    BraveActionsContainer(Browser* browser, Profile* profile);
+    ~BraveActionsContainer() override;
+    void Init();
+    void Update();
+    void SetShouldHide(bool should_hide);
+
+    // ToolbarActionView::Delegate
+    content::WebContents* GetCurrentWebContents() override;
+    bool ShownInsideMenu() const override;
+    // Notifies that a drag completed.
+    void OnToolbarActionViewDragDone() override;
+    // Returns the view of the toolbar actions overflow menu to use as a
+    // reference point for a popup when this view isn't visible.
+    views::MenuButton* GetOverflowReferenceView() override;
+    // Returns the preferred size of the ToolbarActionView.
+    gfx::Size GetToolbarActionSize() override;
+    // Overridden from views::DragController (required by ToolbarActionView::Delegate):
+    void WriteDragDataForView(View* sender,
+                              const gfx::Point& press_pt,
+                              ui::OSExchangeData* data) override;
+    int GetDragOperationsForView(View* sender, const gfx::Point& p) override;
+    bool CanStartDragForView(View* sender,
+                            const gfx::Point& press_pt,
+                            const gfx::Point& p) override;
+
+    // ExtensionRegistryObserver:
+    void OnExtensionLoaded(content::BrowserContext* browser_context,
+                          const extensions::Extension* extension) override;
+    void OnExtensionUnloaded(content::BrowserContext* browser_context,
+                            const extensions::Extension* extension,
+                            extensions::UnloadedExtensionReason reason) override;
+
+    // ExtensionActionAPI::Observer
+    // Called when there is a change to the given |extension_action|.
+    // |web_contents| is the web contents that was affected, and
+    // |browser_context| is the associated BrowserContext. (The latter is
+    // included because ExtensionActionAPI is shared between normal and
+    // incognito contexts, so |browser_context| may not equal
+    // |browser_context_|.)
+    void OnExtensionActionUpdated(
+        ExtensionAction* extension_action,
+        content::WebContents* web_contents,
+        content::BrowserContext* browser_context) override;
+
+  private:
+    void AddShields(const extensions::Extension* brave_extension);
+    void RemoveShields();
+
+    bool should_hide_ = false;
+
+    ToolbarActionView* shields_button_view_ = nullptr;
```

Comment:
How about using `std::unique_ptr` for |shields_button_view_|?  because this class owns it.
In chromium, using raw point means it doesn't own it.

Reply:
I copied this pattern from chromium, for example at https://cs.chromium.org/chromium/src/chrome/browser/ui/views/location_bar/location_bar_view.h?type=cs&q=LocationBarView&sq=package:chromium&g=0&l=456

Could you let me know what the difference is?

Reply:
It is different. Child view you referenced is owned by parent view. That means child view is destroyed when parent view destroyed.
In our case, we delete it explicitly. That means deleting can happen from view tree or our `delete`.

Reply:
Re-reading what chromium does for dynamic extension browser action icons: https://cs.chromium.org/chromium/src/chrome/browser/ui/views/toolbar/browser_actions_container.cc?type=cs&sq=package:chromium&g=0&l=158-179

Will consider not `delete`ing the view manually and maybe just RemoveChildView, or using `shields_button_view_->set_owned_by_client()` as suggested below.

Reply:
Yep, I think `RemoveChildView` is more safer than `delete`.

Reply:
I read again and I think worried situation(double delete) would not be happened.
Explicit deleting is done when `OnExtensionUnloaded()` is called and implicit deleting is done when parent(this class) is destroyed.
When parent is destroyed first, `OnExtensionUnloaded()` cannot be called again.
When `OnExtnsionUnloaded()` is called first, button view is removed from parent and it is explicitly deleted.
So, I think current impl doesn't have problem. WDYT?

Reply:
Yes @simonhong after thinking again I think we should keep the `delete` here since after calling `RemoveChildView` the view will never get `delete`d by the old parent, so we will never free the memory. And we cannot re-use the `ToolbarActionView` for another instance of the extension.